mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp

```diff
@@ -38,29 +38,80 @@ LogicalResult ZneLowering::match(mitigation::ZneOp op) const { return success();
 
 void ZneLowering::rewrite(mitigation::ZneOp op, PatternRewriter &rewriter) const
 {
+    auto point = rewriter.saveInsertionPoint();
     Location loc = op.getLoc();
-
+    auto moduleOp = op->getParentOfType<ModuleOp>();
     // Number of folds
     auto numFolds = op.getNumFolds();
     RankedTensorType numFoldType = cast<RankedTensorType>(numFolds.getType());
     const auto sizeInt = numFoldType.getDimSize(0);
 
     // Folding type
     auto foldingAlgorithm = op.getFolding();
+    auto calleeOp = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, op.getCalleeAttr());
 
-    // Create the folded circuit function
-    FlatSymbolRefAttr foldedCircuitRefAttr =
-        getOrInsertFoldedCircuit(loc, rewriter, op, foldingAlgorithm);
-    func::FuncOp foldedCircuit =
-        SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, foldedCircuitRefAttr);
-
+    TypeRange originalTypes = calleeOp.getArgumentTypes();
+    SmallVector<Type> typesFolded(originalTypes.begin(), originalTypes.end());
+    Type indexType = rewriter.getIndexType();
+    typesFolded.push_back(indexType);
+    func::FuncOp fnFoldedOp;
+    if (!calleeOp->hasAttr("qnode")) {
+        rewriter.setInsertionPointToStart(moduleOp.getBody());
+        FunctionType fnFoldedType = FunctionType::get(op.getContext(), /*inputs=*/
+                                                      typesFolded,
+                                                      /*outputs=*/calleeOp.getResultTypes());
+        std::string fnFoldedName = calleeOp.getName().str() + ".zne";
+        fnFoldedOp = rewriter.create<func::FuncOp>(loc, fnFoldedName, fnFoldedType);
+
+        rewriter.cloneRegionBefore(calleeOp.getBody(), fnFoldedOp.getBody(), fnFoldedOp.end());
+
+        Block *fnFoldedOpBlock = &fnFoldedOp.getBody().front();
+        rewriter.setInsertionPointToStart(fnFoldedOpBlock);
+        fnFoldedOpBlock->addArgument(fnFoldedOp.getArgumentTypes().back(), loc);
+
+        fnFoldedOp.walk([&](func::CallOp callOp) {
```

Comment:
I think this will only work with one level of nesting to get to the qnodes, but I think in principle it is possible that a qnode might be behind two function calls. To solve this one would need to traverse the call graph and look for qnodes, although the downside would be that each traversed function would need duplication as well in case it is used in another context :/

Reply:
Maybe we can start with this, but then we should add verification that qnodes are indeed only present in one layer of nesting to avoid bugs in the future.

Reply:
Yes I will add a check, thanks!

Reply:
Okay so I have added an unoptimized version when they are nested calls.

1. We traverse the call graph and create a `function.zne` for all calls even the classical one that don't contains any qnodes.
2. We traverse the call graph a second time and replace the function call to the `function.zne` equivalent.

Things to improve:
1. Do it in one traversal.
2. Avoid creating `function.zne` for classical functions that do not contains (nested) qnode. This involves traversing the call graph of all classical functions.

Reply:
Thanks, I will take a look! Are you saying that we always clone the entire call graph?